common.go

```diff
@@ -176,11 +177,52 @@ func sum512Truncate256(in []byte) [32]byte {
 	return sliceToByte32(sum512[:32])
 }
 
-func computePayloadHash(headerHash headerHash, nonce Nonce, payloadCiphertext []byte) payloadHash {
+func checkCiphertextState(version Version, ciphertext []byte, isFinal bool) error {
+	makeErr := func() error {
+		return fmt.Errorf("invalid ciphertext state: version=%s, len(ciphertext)=%d, isFinal=%t", version, len(ciphertext), isFinal)
+	}
+
+	switch version.Major {
+	case 1:
+		if len(ciphertext) < secretbox.Overhead {
+			return makeErr()
+		}
+
+		if (len(ciphertext) == secretbox.Overhead) != isFinal {
+			return makeErr()
+		}
+
+		return nil
+	case 2:
+		if len(ciphertext) < secretbox.Overhead {
+			return makeErr()
+		}
+
+		// With v2, it's valid to have a final packet with
+		// non-empty plaintext, so the below is the only
+		// remaining invalid state.
+		if (len(ciphertext) == secretbox.Overhead) && !isFinal {
+			return makeErr()
+		}
```

Comment:
I wonder if we should be stricter, and say that v2 is _never_ allowed to include an empty packet? That assumes we get our "don't packetize at the exactly-1MB-mark until you either get more input or the user calls `Close`" logic correct.